### ISSUE TITLE
Removing watches from a goroutine consuming fsnotify events can lead to a deadlock

### ISSUE THREAD

**Issue by tsuna** [fsnotify/fsnotify#123]
As a result of the fix in fsnotify/fsnotify#73, `watch.Remove()` now synchronously waits for the main fsnotify goroutine (`readEvents()`) to receive and process the `IN_IGNORED` event that the kernel sends to acknowledge the removal of the watch descriptor.  But there could be other inotify events delivered in between, and this can easily cause `readEvents()` to attempt to dispatch an event to someone waiting for `readEvents()` to process the `IN_IGNORED` event, hence the deadlock.

An example of this issue can be found in hpcloud/tail#75 (along with a workaround for that code).

**Comment by purpleidea** [comment]
I really owe a huge thanks to @tsuna for finding this issue. I've spent more time than I am happy to admit realizing that this is what seems to have caused a huge regression in https://github.com/purpleidea/mgmt/. I realized it was the fsnotify code causing the problem, and eventually found this ticket.

I haven't added a workaround to my code yet, and instead will use an older version of this lib for now, but this is a very real problem. I don't completely understand the fix, so patches and hints welcome :)

**Comment by tsuna** [comment]
I can't find a reliable / general way to work around this bug.

* I'm watching a directory `foo` and my goal is to recursively watch the tree under `foo`.
* I delete and re-create a subdirectory `foo/bar` repeatedly in rapid succession.
* After a few iterations, I see a `REMOVE` event being emitted, followed by a `CREATE` event immediately thereafter.
* When my code handles the `REMOVE` event, I need to call `Remove("foo/bar")`, but I can't do so from whichever goroutine is receiving the event from fsnotify due this bug.  So I defer to another goroutine to call `Remove()` for me.
* My code then handles the `CREATE` event and attempts to re-watch this directory by calling `Add("foo/bar")`.

Now we have two goroutines executing `Add("foo/bar")` and `Remove("foo/bar")` concurrently.  Let's say that `Add()` goes in first.  In `Add()` we have a first critical section:

```go
    w.mu.Lock()
    watchEntry, found := w.watches[name]
    w.mu.Unlock()
```

in this scenario, `found` is `true`, because the goroutine calling `Remove()` hasn't gotten around to doing anything yet.  Then we issue the system call without holding the lock, followed by another critical section:

```go
    wd, errno := unix.InotifyAddWatch(w.fd, name, flags)
    if wd == -1 {
        return errno
    }

    w.mu.Lock()
    w.watches[name] = &watch{wd: uint32(wd), flags: flags}
    w.paths[wd] = name
    w.mu.Unlock()
```

Now while I'm in the system call, the `Remove()` goroutine starts to do its thing:

```go
    w.mu.Lock()
    defer w.mu.Unlock()
    watch, ok := w.watches[name]
```

So here `ok` is `true` as well and now `Remove()` will run to completion because it only releases the lock once it's done via the `defer` statement.  So it'll execute:

```go
    success, errno := unix.InotifyRmWatch(w.fd, watch.wd)
```

And then it'll go in this loop that waits for the watch to be removed by the fsnotify goroutine:

```go
    exists := true
    for exists {
        w.cv.Wait()
        _, exists = w.watches[name]
    }
```

When we go in `Wait()`, the lock is released, and `Add()` finishes, overwriting the entry in the maps with a different value for `wd` than before.

Meanwhile fsnotify is trying to deliver the next event (which isn't, yet, the `IN_IGNORED` event required to unblock the `Remove()` call), but it's blocked trying to send on the `w.Events` channel because my code is trying to finish handling the previous `Add()`, which requires sending a message on a channel to the goroutine that is stuck in `Remove()`.


Basically it seems that the only safe way to call `Remove()` is by doing `go watcher.Remove(path)`, but this creates other problems like not knowing exactly when I'm done removing a watch before I add another one.


I'd like to see the code of a reliable recursive watch with fsnotify.  Currently I believe it's not possible to implement due to this bug (and due to #97, but that's another issue).

**Comment by nathany** [comment]
Thanks for investigating this @tsuna. 

Have you looked at related open issues already? Possibly #115.

**Comment by purpleidea** [comment]
@tsuna I have a (possibly broken) implementation of a recursive watcher here: https://github.com/purpleidea/mgmt/blob/master/recwatch/recwatch.go#L19 I did rollback the version of this lib to avoid this bug though. I also think I found at least one issue with my lib anyways. In any case, HTH and a permanent fix is definitely welcome.

**Comment by tsuna** [comment]
@nathany issue #115 and this issue are definitely the same issue.

@purpleidea: I know how to write a recursive watch (and in fact I've run into this bug with the one I wrote), but what I meant is we need some sort of an integration test that stresses the filesystem and implements a recursive watch, to demonstrate that it's feasible to write this with fsnotify.  I assert that currently it is not possible to do this correctly with fsnotify, one will either deadlock, miss events, have watches removed when they shouldn't be, and not deal properly with the inotify queue overflowing in the kernel.

**Comment by nathany** [comment]
What do you think of splitting out inotify as a dependency and adding some inotify-specific stress tests there? #173.

Also, there are some open pull requests in need of review/testing that may help, such as #149.
